### ISSUE TITLE
Fix routing/runoff for coupled model applications

### ISSUE THREAD

**Issue by sanAkel** [GEOS-ESM/GEOSgcm_GridComp#681]
# Statement:
Routing table file (e.g., `CF0180x6C_TM1440xTM1080-Pfafstetter.TRN`) that gets generated by [`make_bcs script(s)`](https://github.com/GEOS-ESM/GEOSgcm_GridComp/blob/develop/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/Raster/make_bcs) puts run off in _wrong_ locations:
- In the middle of ocean (tile) points (some could be islands?).
- On land (tile) points.

See following figure, locations are marked in panels (c) and (d). Attached is a terrain map of the Indonesian seas (W Pac) region that can be compared to panel (d).
These have major consequences for the coupled model: causing it to crash and _leaks_ in the hydrological budget, to name a few.
Comments from @atrayano, @rdkoster and @gmao-rreichle will help.

@yvikhlya has been running [this utility](https://github.com/GEOS-ESM/GMAO_Shared/blob/main/GEOS_Util/coupled_diagnostics/g5lib/src/reroute_runoff/reroute.py) to _fix_ above problem to some extent.

# History:
- @amolod has asked for a fix since _many_ years. 
- More recently effort to _centralize_ boundary conditions (files): @biljanaorescanin and generation (source code) by  @weiyuan-jiang et al., requires us to take a fresh look at it.

# List of steps to generate a routing file that is currently used:
For e.g., `/discover/nobackup/projects/gmao/ssd/aogcm/atmosphere_bcs/Icarus-NLv3/MOM6/CF0180x6C_TM1440xTM1080_newtopo/CF0180x6C_TM1440xTM1080-Pfafstetter.TRN`

1. Run make_bcs **package**. This will produce a TRN file, but it directs runoff to the land, that is incorrect (see below figures).

2. Take the TRN file from above step 1 and run @yvikhlya's `reroute.py` utility (see code at `https://github.com/GEOS-ESM/GMAO_Shared/blob/main/GEOS_Util/coupled_diagnostics/g5lib/src/reroute_runoff/reroute.py`) which fixes the routing so it ends up being in the ocean.

3. Incorporate @lcandre2 's runoff pathways so that runoff from glaciers is routed to the ocean in a more realistic way. **Where is the script/code?**  ETA: I have this code (though is now several years old). It assigns tiles to a 'topographically appropriate' ocean cell. Because the current routing scheme simply puts runoff in the geographically nearest ocean cell (which is incorrect as frequently as it is correct).  The longer term fix  - which needs to be started - is to develop catchments and a pfafstetter routing scheme for landice and treat landice as a true child of surface. 

4. Possibly (if the coupled model or ocean component blows up at run time) iterate the above 2 and/or 3 if too much freshwater is sent to the ocean. This will be done manually by 
modifying the `ocean mask` used in steps 2 and 3, **such _modified_ `ocean mask`, used for routing should be saved somewhere**.

**Note**:
- Step 4 cannot be codified.
- Current [`make_bcs script(s)`](https://github.com/GEOS-ESM/GEOSgcm_GridComp/blob/develop/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/Raster/make_bcs) does not include above steps 2- 4, they are being done outside (manually) as described ⬆️ .

# Action items:
- We need to have steps 2 and 3 included in the `make_bcs script(s)`.

@yvikhlya 's comments on how to get :arrow_up: done:
```
Actually, the problem is in this utility

https://github.com/GEOS-ESM/GEOSgcm_GridComp/blob/develop/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/Raster/mk_runofftbl.F90

What it should do at the very start is to read tile file, read MOM mask and mark all tiles which fall on MOM land as land (change code from 0 to 20). 
Then do what it currently does. This would eliminate re-routing step.

If manual manipulation with routing is necessary, we would feed this utility a fake mask. Done.
```
cc: @wmputman, @tclune, @atrayano, @weiyuan-jiang 

![input_routing_table](https://user-images.githubusercontent.com/18061653/205459006-330e2d69-76bb-4a40-96f3-9efb0bdd3c04.png)

<img width="1663" alt="Screen Shot 2022-12-03 at 2 30 39 PM" src="https://user-images.githubusercontent.com/18061653/205459061-f8b1a514-de3d-4839-963a-683755db5a30.png">

**Comment by sanAkel** [comment]
FWIW, I tried @yvikhlya's [utility](https://github.com/GEOS-ESM/GMAO_Shared/blob/main/GEOS_Util/coupled_diagnostics/g5lib/src/reroute_runoff/reroute.py) myself.
1. It did not reproduce the same output that @yvikhlya has `/discover/nobackup/projects/gmao/ssd/aogcm/atmosphere_bcs/Icarus-NLv3/MOM6/CF0180x6C_TM1440xTM1080_newtopo/CF0180x6C_TM1440xTM1080-Pfafstetter.TRN`
2. However, it seems to _work_ see below.
![try_reroute](https://user-images.githubusercontent.com/18061653/205470548-940dfac0-1824-4a63-8bba-74efad33520e.gif)

**Comment by sanAkel** [comment]
Tagging https://github.com/GEOS-ESM/GEOSgcm_GridComp/issues/495

**Comment by gmao-rreichle** [comment]
@sanAkel 
Thanks for writing this up, very helpful.  But I'm not convinced that the following statement is correct:

> Routing table file (e.g., CF0180x6C_TM1440xTM1080-Pfafstetter.TRN) that gets generated by [make_bcs](https://github.com/GEOS-ESM/GEOSgcm_GridComp/blob/develop/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/Raster/make_bcs) puts run off in wrong locations:

>
>In the middle of ocean (tile) points (some could be islands?).
>On land (tile) points.

If by "make_bcs" here you're referring to the existing utility (that is, only Step 1. in the "List of steps to generate a routing file that is currently used" above), then the (lat/lon) locations that receive the runoff make sense from a land perspective.  The points "in the middle of the ocean" are islands, and runoff from islands should go into the ocean.  One example is Key West (see image below).  Runoff can also seemingly be dumped "on land (tile) points" when estuaries stretch inland.  One such example is in north-western Borneo (see image below). 

I recognize that from the ocean model perspective, it may not work to receive runoff "on land (tile) points" because MOM doesn't operate there, and I thought that this is what Steps 2.-4. are meant to address.  I also agree that it would be good to include Steps 2.-4. in make_bcs.  But I don't necessarily see anything wrong with make_bcs in its existing scope (i.e., **before** the connections to the ocean model are made).

Does that make sense?  I'm mostly still trying to understand the problem. 

<img width="822" alt="Florida" src="https://user-images.githubusercontent.com/54944691/205493200-ecc2bd8d-2dd5-48f8-8cf1-2427b2fdc439.PNG">

<img width="988" alt="Borneo" src="https://user-images.githubusercontent.com/54944691/205493209-fa33edbc-04ee-40d0-8775-44f81934f87c.PNG">

**Comment by sanAkel** [comment]
Thank you @gmao-rreichle. 

> @sanAkel Thanks for writing this up, very helpful. But I'm not convinced that the following statement is correct:
> 
> > Routing table file (e.g., CF0180x6C_TM1440xTM1080-Pfafstetter.TRN) that gets generated by [make_bcs](https://github.com/GEOS-ESM/GEOSgcm_GridComp/blob/develop/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/Raster/make_bcs) puts run off in wrong locations:
> 
> > In the middle of ocean (tile) points (some could be islands?).
> > On land (tile) points.
> 
> If by "make_bcs" here you're referring to the existing utility (that is, only Step 1. in the "List of steps to generate a routing file that is currently used" above), then the (lat/lon) locations that receive the runoff make sense from a land perspective. 

You are correct. I should have been more precise, for instance refer to `mk_rounofftbl.f90`? But I didn't because I am not entirely clear of the `make_bcs` package workflow. Therefore, in the ⬆️ I now refer to `make_bcs script(s)` **as a package.**

>The points "in the middle of the ocean" are islands, and runoff from islands should go into the ocean. One example is Key West (see image below). Runoff can also seemingly be dumped "on land (tile) points" when estuaries stretch inland. One such example is in north-western Borneo (see image below).
> 

Indeed. When the ocean is coarse (fine) resolution this problem may (may not) be so difficult. But from what I can tell, those islands have not (yet?) caused troubles. But should certainly think ...

> I recognize that from the ocean model perspective, it may not work to receive runoff "on land (tile) points" because MOM doesn't operate there, and I thought that this is what Steps 2.-4. are meant to address. I also agree that it would be good to include Steps 2.-4. in make_bcs. 

Since you agree to it as well, would you please take a look at @yvikhlya's script: `https://github.com/GEOS-ESM/GMAO_Shared/blob/main/GEOS_Util/coupled_diagnostics/g5lib/src/reroute_runoff/reroute.py` 

I am certain @atrayano will add to this conversation.

> But I don't necessarily see anything wrong with make_bcs in its existing scope (i.e., **before** the connections to the ocean model are made).
> 
Right! My _mistake_! As said above, I now refer to `make_bcs script(s)` as a whole.

> Does that make sense? I'm mostly still trying to understand the problem.

Very much! 👍 
Thanks again!
> 
> <img alt="Florida" width="822" src="https://user-images.githubusercontent.com/54944691/205493200-ecc2bd8d-2dd5-48f8-8cf1-2427b2fdc439.PNG">
> 
> <img alt="Borneo" width="988" src="https://user-images.githubusercontent.com/54944691/205493209-fa33edbc-04ee-40d0-8775-44f81934f87c.PNG">

**Comment by yvikhlya** [comment]
All, as @sanAkel mentioned above, the problem is in mk_runofftbl.F90, which is called by mk_bcs. It should be aware of MOM land-sea mask and should not put runoff on MOM land. In GEOS set up, MOM land-sea mask is different from GEOS mask and this creates a number of complications in coupled model, including runoff re-routing. If mk_runofftbl.F90 was aware of MOM mask, re-routing step would not be necessary.

**Comment by sanAkel** [comment]
Plots with MOM's land mask (black color; value=0.)

![image](https://user-images.githubusercontent.com/18061653/206331029-a2d94806-5e40-4b67-abc7-2de6aa7e8b86.png)

![image](https://user-images.githubusercontent.com/18061653/206331142-18b2e035-63d5-4e10-8ed5-4877674e7389.png)

![image](https://user-images.githubusercontent.com/18061653/206332474-09097d75-9400-4fd9-a071-764d19c0c0dc.png)

**Comment by atrayano** [comment]
I agree with @yvikhlya - the correct way to "fix" the issue is to make mk_runofftbl.F90 aware of the ocean mask. My only issue is that reading of the ocean mask should be portable, so that it works for MOM (and MOM6), as well as for MITgcm

**Comment by yvikhlya** [comment]
> I agree with @yvikhlya - the correct way to "fix" the issue is to make mk_runofftbl.F90 aware of the ocean mask. My only issue is that reading of the ocean mask should be portable, so that it works for MOM (and MOM6), as well as for MITgcm

Atanas, reading ocean mask IS portable now, since we have this MAPL_Tripolar.nc file regarless which ocean model we use. The mask should be red from this file. Then i,j for ocean tiles should be read from tile file and if mask(i,j)==1, then this tile should be labeled as land.

**Comment by atrayano** [comment]
The is portable for tripolar grids. MITgcm uses lat-lon-cap (LLC) grid. The good news is that they have the same assumption that mask is strictly 0 or 1.  And a minor correction to Yury's comment, the corect type for land is 100, not 20. 20 is the code for land-ice

**Comment by yvikhlya** [comment]
> The is portable for tripolar grids. MITgcm uses lat-lon-cap (LLC) grid. The good news is that they have the same assumption that mask is strictly 0 or 1. And a minor correction to Yury's comment, the corect type for land is 100, not 20. 20 is the code for land-ice

It does not matter if the ocean grid is tripolar or LLC, in any case we'll extract mask from native ocean grid file and put it to MAPL_Tripolar.nc, perhaps the name will need to be changed to MAPL_OceanGrid.nc or something like this.

**Comment by atrayano** [comment]
Yes, we can extract the MIT ocean mask and put in a netCDF file. My preference would be to call it with a more general name, even MAPL seems a bit awkward in naming such file. But the bottom line is that we should read the mask from a file and make few changes to the file mk_runofftabl.F90. I'll try to do this by the end of next week

**Comment by yvikhlya** [comment]
>  But the bottom line is that we should read the mask from a file and make few changes to the file mk_runofftabl.F90. I'll try to do this by the end of next week

Yes, this sounds like a good solution. This will eliminate all this re-route step with f2py and other ugliness.

**Comment by sanAkel** [comment]
> Yes, we can extract the MIT ocean mask and put in a netCDF file. My preference would be to call it with a more general name, even MAPL seems a bit awkward in naming such file. But the bottom line is that we should read the mask from a file and make few changes to the file mk_runofftabl.F90. I'll try to do this by the end of next week

Thank you @atrayano and @yvikhlya for the ⬆️ discussion. Agreed with (almost all of) it (see ⬇️ for exception).

Looking at `MAPL_Tripolar.nc` that is:
- `/discover/nobackup/projects/gmao/ssd/aogcm/ocean_bcs/MOM6/1440x1080_newtopo/MAPL_Tripolar.nc`
It has been (can be) generated from [this code](https://github.com/GEOS-ESM/GEOS_OceanGridComp/tree/develop/MOM6_GEOSPlug/mom6_app/mk_MAPL_Tripolar).

- As the [doc: README.md for MAPL_Tripolar](https://github.com/GEOS-ESM/GEOS_OceanGridComp/blob/develop/MOM6_GEOSPlug/mom6_app/mk_MAPL_Tripolar/README.md) says, it uses data that was generated as part of MOM `INPUT/` files ( to be specific: `/discover/nobackup/projects/gmao/ssd/aogcm/ocean_bcs/MOM6/1440x1080_newtopo/INPUT/`) and _puts_ it in the `MAPL_Tripolar.nc`. 
 
So in essence, it is duplicating information ➡️ copying info from `INPUT/` into `MAPL_Tripolar.nc` This duplication is unnecessary. It's best we to avoid it. **I am not saying we change** `MAPL_Tripolar.nc` **That would be a separate** issue.

But I am not _favoring_ another _copy_ of the mask file:
> It does not matter if the ocean grid is tripolar or LLC, in any case we'll extract mask from a native ocean grid file and put it to MAPL_Tripolar.nc, perhaps the name will need to be changed to MAPL_OceanGrid.nc or something like this.

Variable: `wet` in `/discover/nobackup/projects/gmao/ssd/aogcm/ocean_bcs/MOM6/1440x1080_newtopo/INPUT/ocean_topog.nc` is the _original_ mask, there is no need to **extract and write it to another file**.

@atrayano please point me to `MITgcm` files, I'll find you the equivalent variable (like MOM's `wet`).

```
If (ocean model == "MOM6"):
 use `ocean_topog.nc:wet`
else if (ocean model == "MITgcm"):
 use `??:??` ⬅️ @atrayano 
else
 **ERROR**
endif
```

**Comment by yvikhlya** [comment]
@sanAkel With MAPL_Tripolar.nc we are avoiding having multiple tripolar grid factories in MAPL, make_bcs, plotting package and any other package which uses the ocean grid. We have 4 different tripolar grid formats and MIT grid is coming. We discussed this many times already.

**Comment by sanAkel** [comment]
> **I am not saying we change** `MAPL_Tripolar.nc` **That would be a separate** issue.

Understood and already said ⬆️ 
>I am not saying we change MAPL_Tripolar.nc That would be a separate issue.

**Comment by lcandre2** [comment]
Just for the record on point 3 in original, @lcandre2 has the topographies, drainage basins, and  code (though is now several years old). It assigns tiles to a 'topographically appropriate' ocean cell. So, my understanding of the order is 1. land tiles are routed to the appropriate river outlet location on the dataocean grid cell. 2. my scripts modify the routing table generated in 1 for necessary landice tiles. 3. Yury's utility then moves the outlets to the nearest model ocean grid. 4. Manually modify model ocean outlet points to deal with model ocean diffusion issues. 

@yvikhlya let me know if I made mistakes.


Because the current routing scheme simply puts runoff in the geographically nearest ocean cell (which is incorrect as frequently as it is correct). The longer term fix - which needs to be started - is to develop catchments and a pfafstetter routing scheme for landice and treat landice as a true child of surface.

**Comment by lcandre2** [comment]
@yvikhlya, agreed. I think the goal now is that landice modifications will be incorporated into the river outlet file and then the mk_runofftbl.F90 will incorporate information about different ocean masks. Will there still need to be be manual adjustments to deal with diffusion issues?

**Comment by yvikhlya** [comment]
> @yvikhlya, agreed. I think the goal now is that landice modifications will be incorporated into the river outlet file and then the mk_runofftbl.F90 will incorporate information about different ocean masks. Will there still need to be be manual adjustments to deal with diffusion issues?

We don't this yet, but if manual adjustments will be necessary, this can be easily done by feeding modified ocean mask to mk_runofftbl.F90

**Comment by sanAkel** [comment]
> Will there still need to be be manual adjustments to deal with diffusion issues?

As @yvikhlya said :arrow_up @lcandre2, we do not know-- until we try. But FWIW, we have not yet run into an issue with MOM6 that needed an adjustment. Collectively we have > 40 years of simulations thus far. (Not having run into an issue does not mean, there won't be one in future! 😮 )

**Comment by gmao-rreichle** [comment]
A quick update: I spoke with @rdkoster and @biljanaorescanin about updating the land information that lists the lat/lon coordinates of the river outlets into the ocean, along with the contributing tile IDs.  This effort will update the original 2-by-2.5-degree information that Max put together with input from @rdkoster and that is currently processed by `make_bcs` into the routing table file (e.g., `/discover/nobackup/projects/gmao/ssd/aogcm/atmosphere_bcs/Icarus-NLv3/MOM6/CF0180x6C_TM1440xTM1080_newtopo/CF0180x6C_TM1440xTM1080-Pfafstetter.TRN`)  (see comments above).
Yujin Zeng (@zyj8881357) will tackle this task, but he is not yet on the NASA IT network (ie, does not have access to NCCS/Discover) and he is not yet a member of the GEOS-ESM team.  But Yujin should be able to read and comment on this Github issue to facilitate information exchange between the land and ocean groups.  
cc: @yvikhlya @sanAkel @lcandre2

**Comment by zyj8881357** [comment]
I will review all this information.

**Comment by sanAkel** [comment]
> I will review all this information.

Most likely you don't have access to MS Teams? In which case, if needed, I can set up a meeting via google meet. Let me know.

**Comment by sanAkel** [comment]
> A quick update: I spoke with @rdkoster and @biljanaorescanin about updating the land information that lists the lat/lon coordinates of the river outlets into the ocean, along with the contributing tile IDs.  This effort will update the original 2-by-2.5-degree information that Max put together with input from @rdkoster and that is currently processed by `make_bcs` into the routing table file (e.g., `/discover/nobackup/projects/gmao/ssd/aogcm/atmosphere_bcs/Icarus-NLv3/MOM6/CF0180x6C_TM1440xTM1080_newtopo/CF0180x6C_TM1440xTM1080-Pfafstetter.TRN`)  (see comments above).
> 
> Yujin Zeng (@zyj8881357) will tackle this task, but he is not yet on the NASA IT network (ie, does not have access to NCCS/Discover) and he is not yet a member of the GEOS-ESM team.  But Yujin should be able to read and comment on this Github issue to facilitate information exchange between the land and ocean groups.  
> 
> cc: @yvikhlya @sanAkel @lcandre2 

Thank you @gmao-rreichle. Appreciate it!

**Comment by zyj8881357** [comment]
> > I will review all this information.
> 
> Most likely you don't have access to MS Teams? In which case, if needed, I can set up a meeting via google meet. Let me know.

I am able to join MS Teams meeting without any problems. Thanks for asking!

**Comment by rdkoster** [comment]
Looking at my old directories, I see that I am the one who originally created (long ago!) the files lonraster.dat and latraster.dat.  Atanas, is it my understanding that if these files are regenerated with the improved routing network information and providing improved ocean outlet information, all of the codes downstream will work?  That is, if you get a new version of lonraster.dat and latraster.dat with identical formatting (and thus identical resolution for now) but improved information content, our job on the land side would be done?

**Comment by atrayano** [comment]
Randy, this seems to be a very reasonable assumption (except perhaps the resolution)

Atanas

From: Randy Koster ***@***.***>
Reply-To: GEOS-ESM/GEOSgcm_GridComp ***@***.***>
Date: Friday, December 16, 2022 at 2:55 PM
To: GEOS-ESM/GEOSgcm_GridComp ***@***.***>
Cc: "Trayanov, Atanas L. (GSFC-610.1)[SCIENCE SYSTEMS AND APPLICATIONS INC]" ***@***.***>, Assign ***@***.***>
Subject: [EXTERNAL] Re: [GEOS-ESM/GEOSgcm_GridComp] Fix routing/runoff for coupled model applications (Issue #681)


Looking at my old directories, I see that I am the one who originally created (long ago!) the files lonraster.dat and latraster.dat. Atanas, is it my understanding that if these files are regenerated with the improved routing network information and providing improved ocean outlet information, all of the codes downstream will work? That is, if you get a new version of lonraster.dat and latraster.dat with identical formatting (and thus identical resolution for now) but improved information content, our job on the land side would be done?

**Comment by lcandre2** [comment]
@rdkoster @atrayano , I agree, but I would like to check/modify the information for landice tiles too.

**Comment by gmao-rreichle** [comment]
@rdkoster, @atrayano, 

Re. lonraster.dat and latraster.dat:  I found copies of these files in land01.gsfc.nasa.gov:/land/koster/ where they were sitting among a ton of other, presumably unrelated stuff and without any documentation.   Are these files archived somewhere with the bcs input data?  I could not find them on Discover.  @biljanaorescanin, can you please capture these files in the bcs_shared project disk and make sure they're documented. 

I guess the following reader provides some information?
`GEOSgcm_GridComp/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/Raster/read_riveroutlet.F90`
If I'm guessing correctly, the files contain 8640-by-4320 longitude and latitude values.  These values indicate for each of the 8640-by-4320 grid cells the longitude and latitude of the river outlet into the ocean.  (How about endorheic basins?) Maybe someone can visualize the content of the current files. 

If that's all correct, then I do **not** think that reproducing these files will give us what we need.  Isn't our current raster grid for tile IDs (based on SRTM data) at much finer resolution than 8640-by-4320?  Wouldn't we want the mapping be from land (and landice) tiles to lat/lon of river outlets?  Or, perhaps alternatively, from the high-res raster file to lat/lon of river outlets?  (I suppose the latter option would not be suitable as a direct input to the model and would, in any case, need to be processed into the former?)

PS: I removed the extraneous email trail from the previous comment.  My recommendation would be to enter comments/replies directly on Github.com and not email-reply to system-generated emails, which gunks up the Github pages.

**Comment by rdkoster** [comment]
The way the old code worked, the coordinates of the final river outlet (to ocean) was identified for each land pixel based on the land mask information we were using.  We then "spiraled out" from those outlet coordinates until we found the closest location considered by the ocean model to be "ocean".  Accordingly, for the new files, we would need to work with the latest ocean mask.  Biljana, can you locate that file and send it to Yujin?  We worked with an 8640x4320 array before (again, a long time ago!), with 0 values in the array indicating ocean points.

**Comment by rdkoster** [comment]
Sorry, I posted my comment before the page refreshed with Rolf's comment.  In any case, I understand how the two files were created and what they contain, and I can guide the creation of the new files at whatever resolution is needed.

**Comment by gmao-rreichle** [comment]
@rdkoster, thanks for the additional info.  Maybe I'm misunderstanding all of this, but I thought what we need to produce is a list of lat/lon coordinates with (final) river outlets (to ocean) based solely on the **land** (and landice) tile space, along with the tile IDs of all the upstream land tiles that contribute to each outlet.  This is what's encoded in the SRTM-based Pfafstetter information.  Once we have the lat/lons of the river outlets and the upstream tile IDs, the ocean group can take a look at the lat/lons and decide which ocean tile(s) should receive the freshwater associated with our river outlets.  
In other words, we (the land group) would **not** be concerned with the ocean mask.  This will make the processing more modular.  There are several land tile spaces (resolutions) and many ocean masks of interest, and the first step (of identifying the river outlets and contributing land(ice) tiles according to the Pfafstetter classification should not depend on the ocean mask.

**Comment by rdkoster** [comment]
I was just describing what went into the previous incarnation of those two files.  The file loncenter.dat, which is currently being used, contains an 8640x4320 array of ocean outlet i-indices (for a 2.5'x2,5' grid) for each 2.5'x2.5' location on the Earth; latcenter.dat contains the equivalent array of j-indices.  These outlet indices were adjusted as described above to agree with old-ocean-mask ocean locations before they were written out.  Anyway, that's the format of the files currently being used to generate the land-to-ocean transfers and, specifically, to key (by some other program, written by Max or Atanas?) the tile-generated runoffs to ocean sinks.  Just trying to explain these two datasets...

**Comment by sanAkel** [comment]
Hi All (@rdkoster, @gmao-rreichle),
- I don't know what was done some 15- or 20- years ago to generate ocean ancillary files (bathymetry, masks, etc), but knowing what we have with MOM5 and now MOM6 based models, much of it was taken from elsewhere-- the key point is that: **there is no traceability** and certainly no "self-control" for content(s) of any dataset. This is all rapidly changing! All I am trying to say is that, it would be easier to say:
 - What do we want to do?
 - What utilities/information (SRTM) do we have?
 - How well does this solve our problem?
 - There on, CI takes over!

(In that spirit, I can assure that none of us can track down the source of `SEAWIFS_KPAR_mon_clim.180x1080`, in `/discover/nobackup/projects/gmao/share/gmao_ops/fvInput/g5gcm/bcs/realtime/OSTIA_REYNOLDS/180x1080/` 😱 )

**Comment by sanAkel** [comment]
## ⬆️ Suggested [fix](https://github.com/GEOS-ESM/GEOSgcm_GridComp/issues/681#issuecomment-1344657012)
happening via https://github.com/GEOS-ESM/GEOSgcm_GridComp/tree/feature/sanAkel/partial-fix-issue-681

**Comment by yvikhlya** [comment]
> ## ⬆️ Suggested fix:
> 
> For [`mk_runofftbl.F90`](https://github.com/GEOS-ESM/GEOSgcm_GridComp/issues/681#issuecomment-1344657012) happening via https://github.com/GEOS-ESM/GEOSgcm_GridComp/tree/feature/sanAkel/partial-fix-issue-681

This `reroute` subroutine is not actually needed. After ocean mask is read, all tile types where `mask(i2, j2)=0`, where `i2, j2` - indexes of second grid in the tile file, should be changed from 0 to 100. This is it.

**Comment by sanAkel** [comment]
Thank you @yvikhlya I was trying to have following discussion with @atrayano:

I wanted to get https://github.com/GEOS-ESM/GEOSgcm_GridComp/issues/681

out of our way. So started this branch: https://github.com/GEOS-ESM/GEOSgcm_GridComp/tree/feature/sanAkel/partial-fix-issue-681

But as I got "closer" to "finishing it" realized that we were wrong:

Thought was that `mk_runofftbl.f90` needs to be "aware" of ocean and its mask. That's what I tried to do ... as @atrayano and @yvikhlya  said.

For that to happen, whatever tiles are inconsistent between .til file (`type is = 0`: ocean) should be "cross-checked" with ocean-mask and if the mask says no, they are not ocean, then *correct* them to be land tile-type, i.e., 100. <-- that's what you concurred with @yvikhlya  ("I agree with @yvikhlya  - the correct way to "fix" the issue is to make `mk_runofftbl.F90` aware of the ocean mask.") https://github.com/GEOS-ESM/GEOSgcm_GridComp/issues/681#issuecomment-1344657012

**But here is the real question**:

- Why did the .til file even make that mistake: i.e., mark those points ocean??
- mk_runtbl.f90 is being handed off a WRONG tile file-- for coupled model purpose. **That's the central problem here- as I see it now**.

Given this, I am not supportive of "patching" a downstream app such as mk_runtbl.f90

FWIW, we can just "package" @yvikhlya's script (I put link in the issue: https://github.com/GEOS-ESM/GMAO_Shared/blob/main/GEOS_Util/coupled_diagnostics/g5lib/src/reroute_runoff/reroute.py) and do a **PROPER** fix. 

**That will really CUT down if not, eliminate other bad issues such as "orphan" points, sink holes in our budgets and whatnot in our COUPLED model.**



It is my understanding that @GEOS-ESM/land-team are not insistent on anything that ends up used for coupled applications: with the ocean. If any, they do want to clean up as well.

**Comment by yvikhlya** [comment]
I support approach suggested by @sanAkel, i.e. moving a fix for river routing from routing table level to tile file level, in other words, making a coupled tile file mask to be identical to ocean model mask. This would solve multiple issues in coupled  model at once. If there are any obstacles in implementing this approach, I would like to know what are these obstacles exactly.

**Comment by atrayano** [comment]
@yvikhlya and @sanAkel Making a .til for the coupled model based on the ocean mask is not a very good idea: this effectively will "modify" the land tile definitions and we will have totally incompatible boundary conditions and restarts between coupled and AMIP runs. I firmly believe that the best compromise for us is to mk_runofftabl.F90 aware of the actual ocean mask and always send the runoff fresh water to an actual ocean point

**Comment by sanAkel** [comment]
@atrayano 

>we will have totally incompatible boundary conditions and restarts between coupled and AMIP runs.

It is my understanding from @biljanaorescanin that we already have **different** boundary conditions between coupled and AMIP. Moreover, it makes sense that they are different! Every run of `make_bcs` generates a totally self-contained set of files, so there is no `mix-and-match` of boundary conditions generated for coupled getting used in AMIP or other way around.

As for:
>>restarts between coupled and AMIP runs.
This is also an existing _pre-condition_, i.e., it already exists! I routine `remap` restarts from AMIP to coupled.

Is it possible to generate such a tile file and **SEE** how it looks?

**Comment by yvikhlya** [comment]
> Making a .til for the coupled model based on the ocean mask is not a very good idea: this effectively will "modify" the land tile definitions and we will have totally incompatible boundary conditions and restarts between coupled and AMIP runs.

We already have different set ob BCs and restarts for AMIP and coupled models. The question is, if we convert some ocean tiles into land tiles, will it be possible to generate BCs and ICs for these tiles? I guess, this is a question for @gmao-rreichle  and @biljanaorescanin

**Comment by gmao-rreichle** [comment]
A couple of thoughts on this:

1. Ideally, the (land/landice/lake/ocean) tile spaces and BCs would be the same for the AGCM and the coupled model.  This isn't the case now.  I'm wondering if there's any way this could be accomplished.  Perhaps we'd need a new type of "estuary" tile that is used to connect the land and the ocean.  Or maybe "lake" tiles could even be used (rather than introducing a new type), given that "lakes" are just infinite water reservoirs with a surface energy budget (presumably what "estuaries" would be initially).  If this can work, then we should be able to interchange BCs and ICs seamlessly between the AGCM (ie, FP & AMIP) and the coupled model.
 
2. Currently, the land tiles and files of the coupled model BCs can be used in the GEOSldas land-only (offline) system, so we can create reasonably spun-up ICs offline.  This isn't optimal, though. Even worse is the "regrid.pl" / "remap_restarts.py" approach of simply interpolating from one (land) tile space to another (which seems to be what @sanAkel is doing to get by). 
 
3. Having said that, if we are indeed making land out of ocean (where ocean is defined by the 10 arcsec mask file GEOS5_10arcsec_mask.nc), then how are soil, topography, vegetation parameters assigned?  I assume this is happening somehow in the current make_bcs post-processing.  Are the parameter values from nearby land tiles duplicated?  Or are the "new" land tiles merged into nearby existing land tiles?  

I can't say I understand what the ocean model considers ocean and how that relates to the reference, 10 arcsec mask file (GEOS5_10arcsec_mask.nc), and I'm probably missing a bunch of things.  But the more I think about what is done to make the coupled model work, the less sense it makes to me. 

cc: @rdkoster @yvikhlya @lcandre2 @biljanaorescanin

**Comment by sanAkel** [comment]
@gmao-rreichle Those are great comments.

I'm finding increasingly convinced that we need to take a fresh/clean look at all this...

As time permits, in next few days I will upload plots of
10 GEOS arc second dataset and GEBCO 15 arc Sec that we use for ocean- that's probably the "other end point"!

**Comment by sanAkel** [comment]
> > Making a .til for the coupled model based on the ocean mask is not a very good idea: this effectively will "modify" the land tile definitions and we will have totally incompatible boundary conditions and restarts between coupled and AMIP runs.
> 
> 
> 
> We already have different set of BCs and restarts for AMIP and coupled models. The question is, if we convert some ocean tiles into land tiles, to make GEOS land-sea mask to be identical to MOM land-sea mask, will it be possible to generate BCs and ICs for these tiles? I guess, this is a question for @gmao-rreichle  and @biljanaorescanin 

Based on @gmao-rreichle's comments, it seems possible. Did I get that right?

And yes, I use regrid.pl or it's python version: remap_restarts.py to convert from one kind of tiles (amip) to another (coupled): I do account for spin up in all my "science" results.

**Comment by yvikhlya** [comment]
> I can't say I understand what the ocean model considers ocean

Ocean model has it own land-sea mask which is not very realistic in some regions. This creates a number of complications in coupling ocean with atmophere/land models.

**Comment by sanAkel** [comment]
> > I can't say I understand what the ocean model considers ocean
> 
> 
> 
> Ocean model has its own land-sea mask which is not very realistic in some regions. This creates a number of complications in coupling ocean with atmophere/land models.
> 
> 
> 
> 

I would say coastline is rather "boxed up" or choppy, not as realistic as "nature".
Well, that gets "better" at higher resolutions. But all that besides the main point of tile file reconciliation.

**Comment by yvikhlya** [comment]
The idea of introducing a new tile type looks interesting.

**Comment by sanAkel** [comment]
With no action, this issue has gone _stale_! The root cause of the problem continues to persist, without any resolution. There is no point in having open issues with no solution.

**Comment by gmao-rreichle** [comment]
@sanAkel: "Stale" isn't quite right.  Yujin has delivered the new, higher-res outlet files to Atanas who was going to use them to create updated TRN files.  Alas, the latter hasn't happened yet.  We just asked Yujin to look into this step.  It may be too big of an ask, given that Yujin has virtually no experience with GEOS.  But he's working on the land routing and needs to understand the TRN file anyway.  With any luck and - hopefully - some support from Atanas, this is still moving ahead.

**Comment by sanAkel** [comment]
Thank you @gmao-rreichle.

I'm eternally hopeful! :)

**Comment by zyj8881357** [comment]
Some updates on my side about this issue: 

The higher-res outlet files I sent to Atanas show the final outlet locations (for each cell) in lon and lat indexes (not degrees) on a global 1-arcmin raster grid. All the outlet locations are on land, so they must be linked to ocean points for the use in the coupled model.  

As I checked the data and mask information, the problem is that there may some "lakes" (water bodies) laying between the land and ocean. The runoff should go through these "lakes" to ocean. If some of these "lakes" are not explicitly simulated in the ocean model, it will be not easy to link the outlet on land to a correct ocean point (for example, simply linking the nearest ocean point to the outlet on land).  

But the good thing is that, for all the ~22000 ocean outlets in the world, most of them only deliver a small amount of runoff to ocean. In these cases, a small shift of the outlet location in ocean will not be a big problem I guess? So we may use the nearest ocean point as the outlet in ocean for these cases. For all other outlets (mean discharge > 100 m3/s, considering that the first 150 major rivers in the world have a mean discharge >= 2000 m3/s), I can calculate the distances between the outlet on land and the nearest ocean point. If the distance is very small (<300 m), it is also OK for linking them to the nearest ocean point? If the distance is not small, I can do manually to link this outlet to an ocean point.
As I checked, only ~60 outlets needs to be set manually. Not a big deal.

So if I have a land-ocean mask used in the ocean model, I can do all this above. But as I learnt from the former posts,  the land-ocean mask can be changed when we change the ocean model (MOM, MIT). So a better way may be that, on my side, I use some high-res mask file (eg. the 10s GEOS land-lake-ocean mask file) to determine the lat and lon degrees (in ocean) for each outlets with the methods above. And then others can link this degree information to ocean tiles very easily (for example, using the nearest ocean tile) no matter what land-ocean mask is in use.

**Comment by yvikhlya** [comment]
Each ocean grid has its own land/sea mask and there are no lakes on any of ocean grids. Ocean mask (MOM/MIT) is not the same as GEOS or realistic mask. Runoff should be routed from land tile into ocean tile which corresponds to wet grid cell in ocean model.

**Comment by zyj8881357** [comment]
> Each ocean grid has its own land/sea mask and there are no lakes on any of ocean grids. Ocean mask (MOM/MIT) is not the same as GEOS or realistic mask. Runoff should be routed from land tile into ocean tile which corresponds to wet grid cell in ocean model.

Thanks for the quick response! So the best way may be that, someone send me the ocean masks for the MOM and MIT, and I work on those masks and send the outlets locations (in ocean) for each of the mask. Then this outlet information can be directly applied in making bcs. I cannot access to NASA computers, so someone needs to send me the mask files...

**Comment by atrayano** [comment]
@zyj8881357 Thanks for your long update. The way I see it, we need to produce ocean mask file(s), and typically those vary by the ocean model and the ocean resolution. I like your idea to "locale" the nearest ocean outlet point using a high-res definition. This will be an improved version of the current scheme. But you do this process only once. Then we run mk_runofftbl.x (which I am supposed to modify and read/process the relevant ocean mask)

**Comment by zyj8881357** [comment]
> @zyj8881357 Thanks for your long update. The way I see it, we need to produce ocean mask file(s), and typically those vary by the ocean model and the ocean resolution. I like your idea to "locale" the nearest ocean outlet point using a high-res definition. This will be an improved version of the current scheme. But you do this process only once. Then we run mk_runofftbl.x (which I am supposed to modify and read/process the relevant ocean mask)

Yes. In a short words, I can provide you files with outlet locations in lat and lon degrees (rather than in indexes). And this version should be very easy to link to the ocean tiles in ocean models.

**Comment by zyj8881357** [comment]
> Let me check the assumptions in the current version of the code. I am nearly 100% sure that it expects that lat-lon info (either in degrees or as index) points to ocean cell (not land cell)

Yes, in the current version (that I mentioned above, not the version I already sent to you), the lat-lon info points to ocean (not land). The ocean is defined in the mask file of "GEOS5_10s_mask.nc". There may still some difference for the ocean mask between "GEOS5_10s_mask.nc" and ocean model (MOM/MIT), but the (ocean->ocean shift) will be much easier to handle than before (land->ocean shift). If you think this way is OK for you, I will go to proceed this updated version and send it to you.

**Comment by sanAkel** [comment]
@zyj8881357 thank you!

Can you upload an ascii file with:


| River (name) | lat (deg N) | lon (deg E) |
| --- | --- | --- | 
| ABCD | y_1 | x_1 |

You should be able to upload it _here_, if not, please share via google drive; assuming no access to discover cluster at NCCS.

Thanks again.

**Comment by zyj8881357** [comment]
> @zyj8881357 thank you!
> 
> Can you upload an ascii file with:
> River (name) 	lat (deg N) 	lon (deg E)
> ABCD 	y_1 	x_1
> 
> You should be able to upload it _here_, if not, please share via google drive; assuming no access to discover cluster at NCCS.
> 
> Thanks again.

I uploaded it to the google drive. The link is:
https://docs.google.com/spreadsheets/d/1FbQsqoYL-knSmF_qzKlOcC8Pp3kxW4Ok/edit?usp=share_link&ouid=110154394974887000797&rtpof=true&sd=true

It's a spreadsheet. I also added the mean discharge for each river. Please check whether these locations are consistent with the ocean mask in the ocean model.

**Comment by sanAkel** [comment]
Thanks @zyj8881357 I downloaded the file. (You can remove it from above URL if you wish to.)

**Question**:
Is `river(catchID)` unique? In other words, does it change with tile file resolution?

**Comment by zyj8881357** [comment]
> Thanks @zyj8881357 I downloaded the file. (You can remove it from above URL if you wish to.)
> 
> **Question**: Is `river(catchID)` unique? In other words, does it change with tile file resolution?

Thanks for the downloading! The catchment IDs are unique and do not change with the tile file resolution. In other words, the outlets locations in the table do not change with the tile resolution.

**Comment by sanAkel** [comment]
> Thanks for the downloading! The catchment IDs are unique and do not change with the tile file resolution. In other words, the outlets locations in the table do not change with the tile resolution.

In that case, 

1. Say the resolution of the atmosphere is `c12`, which is _very_ coarse, how do you propose to deal with more than one catchment ID having the same (lon, lat)?

2. See attached figure/notebook, what happened to Greenland? You don't include glaciers?
<img width="572" alt="Screen Shot 2023-02-07 at 10 39 22 PM" src="https://user-images.githubusercontent.com/18061653/217424650-ac975fd2-a73b-44e4-90d6-e0d91f9de49e.png">

**Comment by gmao-rreichle** [comment]
@sanAkel:

> Say the resolution of the atmosphere is c12, which is very coarse, how do you propose to deal with more than one catchment ID having the same (lon, lat)?

I presume this would have to be handled in the mk_runofftbl.F90, and it's not different from what would have been needed with the older outlet info.  What Yujin provides is the best location of the outlet based on the currently used Pfafstetter basin/watershed classification, which has a fixed "resolution".   

> [...] what happened to Greenland? You don't include glaciers?

Yujin only provides the outlets from the land.  I don't know how exactly @lcandre2 handles glacier runoff from Greenland.  There should be some (non-glaciated) land along the coast, especially in the north of Greenland.  In your map above, I don't see outlets from Greenlands "land".  That's something for Yujin to look into.  Possibly Yujin and Lauren left **all** of Greenland to Lauren's processing?

**Comment by zyj8881357** [comment]
> > Thanks for the downloading! The catchment IDs are unique and do not change with the tile file resolution. In other words, the outlets locations in the table do not change with the tile resolution.
> 
> In that case,
> 
>     1. Say the resolution of the atmosphere is `c12`, which is _very_ coarse, how do you propose to deal with more than one catchment ID having the same (lon, lat)?
> 
>     2. See attached figure/notebook, what happened to Greenland? You don't include glaciers?
> 
> 
> <img alt="Screen Shot 2023-02-07 at 10 39 22 PM" width="572" src="https://user-images.githubusercontent.com/18061653/217424650-ac975fd2-a73b-44e4-90d6-e0d91f9de49e.png">

Thanks for the feedback! Now I added three outlets for the Greenland (as shown by the attached figure). These three outlets are for the land area (non-glaciated) in the Greenland (as they have river IDs: 154779, 154795 and 161496).  I did not treat the glacier area because we don't have the river network information (ie, Pfaf code) associated with the glacier area. You can access the updated version of the outlet locations at:
https://docs.google.com/spreadsheets/d/1rOG4KhQLbP4YJKYhqP6fqIkHoG3wjz5w/edit?usp=share_link&ouid=110154394974887000797&rtpof=true&sd=true

As mentioned by @gmao-rreichle, this table only lists the locations of all outlets to ocean. It cannot be directly used in the code (ie, read_riveroutlet.F90 and mk_runofftbl.F90) for making bcs. At this stage, please just check if all this points are not difficult for you to transfer to the ocean area in your ocean models. If some of them (especially for those with mean discharge > 100 m3/s) are not easy for you to transfer to your ocean area, just tell me their river IDs and show me the land-ocean mask you are using. So I can fix it further. After all this outlet locations are OK with you to use, I will go to proceed them to the format that can be easily read in the read_riveroutlet.F90 and mk_runofftbl.F90. Thank you!

<img width="738" alt="image" src="https://user-images.githubusercontent.com/32678746/218155526-24cc3895-04c6-445c-aa77-64ccd399d78b.png">

**Comment by gmao-rreichle** [comment]
Thanks for the update, @zyj8881357.  Re. Greenland, I wasn't sure if we have Pfafstetter info for the non-glaciated land area.  It sounds like we do not.  Are the three outlets you added for Greenland meant to collect runoff from all of the non-glaciated land in Greenland?  This doesn't seem right to me.  Shouldn't there be a much larger number of outlets surrounding the entire coast?  I suppose this should be looked at in conjunction with how @lcandre2 routes the glacier runoff towards the coast.

**Comment by lcandre2** [comment]
Thanks for the tag @gmao-rreichle. @zyj8881357, we do not currently have pfastetter codes for Greenland or Antarctica because most runoff flows through the ice sheet and then along the ice sheet bed to a land or ocean outlet. I have a routing code that addresses this and provides better output locations for greenland and antarctica (in the rare cases it has it) runoff. However, it is offline.  @zyj8881357, can we touch base via a teams meeting to coordinate how to approach this?

**Comment by zyj8881357** [comment]
> Thanks for the update, @zyj8881357. Re. Greenland, I wasn't sure if we have Pfafstetter info for the non-glaciated land area. It sounds like we do not. Are the three outlets you added for Greenland meant to collect runoff from all of the non-glaciated land in Greenland? This doesn't seem right to me. Shouldn't there be a much larger number of outlets surrounding the entire coast? I suppose this should be looked at in conjunction with how @lcandre2 routes the glacier runoff towards the coast.

If you look at the figure, the green area is for the glacier area and there is no Pfafstetter code associated to it. In our GEOS5_10arcsec_mask.nc file, there are only three unit catchments around it. Their Pfafstetter code tells us they are isolated catchments without any upstream and downstream catchments. I agree with you that our Pfafstetter code is too simple over the Greenland, and it would be better for us to treat the Greenland in conjunction with glacier people.

**Comment by lcandre2** [comment]
> > Thanks for the update, @zyj8881357. Re. Greenland, I wasn't sure if we have Pfafstetter info for the non-glaciated land area. It sounds like we do not. Are the three outlets you added for Greenland meant to collect runoff from all of the non-glaciated land in Greenland? This doesn't seem right to me. Shouldn't there be a much larger number of outlets surrounding the entire coast? I suppose this should be looked at in conjunction with how @lcandre2 routes the glacier runoff towards the coast.
> 
> If you look at the figure, the green area is for the glacier area and there is no Pfafstetter code associated to it. In our GEOS5_10arcsec_mask.nc file, there are only three unit catchments around it. Their Pfafstetter code tells us they are isolated catchments without any upstream and downstream catchments. I agree with you that our Pfafstetter code is too simple over the Greenland, and it would be better for us to treat the Greenland in conjunction with glacier people.

@zyj8881357 We must have been writing at the same time! see my message above. Let's chat.

**Comment by zyj8881357** [comment]
> Thanks for the tag @gmao-rreichle. @zyj8881357, we do not currently have pfastetter codes for Greenland or Antarctica because most runoff flows through the ice sheet and then along the ice sheet bed to a land or ocean outlet. I have a routing code that addresses this and provides better output locations for greenland and antarctica (in the rare cases it has it) runoff. However, it is offline. @zyj8881357, can we touch base via a teams meeting to coordinate how to approach this? It might be the right time to develop catchments for the ice sheets. We have the data, I just need to do the work.

Yes. I just saw it! We can set up a teems meeting to see how to coordinate the land and glacier. If you think it is better to discuss the meeting schedule via Email, you email me at yujinz@umbc.edu. Thank you!

**Comment by sanAkel** [comment]
@zyj8881357, Can I ask you for a help?

Could you make a `diff` plot of ALL land data from [here](https://portal.opentopography.org/datasetMetadata?otCollectionID=OT.122019.4326.1) which has SRTM?

To be precise, plot of `land data from GEOS5_10s_mask.nc` - ⬆️ URL. If the land difference is zero or what @GEOS-ESM/land-team think are sensible, for once, we can switch to a consistent source of topography for entire earth! 🚀

**Comment by zyj8881357** [comment]
> @zyj8881357, Can I ask you for a help?
> 
> Could you make a `diff` plot of ALL land data from [here](https://portal.opentopography.org/datasetMetadata?otCollectionID=OT.122019.4326.1) which has SRTM?
> 
> To be precise, plot of `land data from GEOS5_10s_mask.nc` - ⬆️ URL. If the land difference is zero or what @GEOS-ESM/land-team think are sensible, for once, we can switch to a consistent source of topography for entire earth! 🚀
> 
> (of course, we can **try/test** first!)

Hi Santha, I think you want me to compare the elevation diff between our land source data and the DEM from the link? I'd like to do the comparison, but the GEOS5_10s_mask.nc only provides the land-sea mask and does not include the elevation values. So I may need to get the elevation source data in our land model first before doing the comparison. @gmao-rreichle @biljanaorescanin

**Comment by sanAkel** [comment]
> Hi Santha, I think you want me to compare the elevation diff between our land source data and the DEM from the link? I'd like to do the comparison, but the GEOS5_10s_mask.nc only provides the land-sea mask and does not include the elevation values. So I may need to get the elevation source data in our land model first before doing the comparison. @gmao-rreichle @biljanaorescanin

@zyj8881357 I recall, sometime ago when I plotted `GEOS5_10arcsec_mask_freshwater-lakes.nc` 
https://github.com/GEOS-ESM/GEOSgcm_GridComp/blob/5938f19b7acfb46ea44ad42f421448ee03762241/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/Raster/makebcs/make_bcs#L437
it had elevation over land and none over the open ocean. Maybe I'm wrong (I do forget things)! But, yes, elevation diff is what I'm looking for ...

I'll also look at `${GLOBAL_CATCH_DATA}` that I'm feeding to the `make_bcs`. Information [such as this](https://portal.opentopography.org/datasetMetadata?otCollectionID=OT.122019.4326.1) has made me curious about any differences!

Thanks!

**Comment by gmao-rreichle** [comment]
@sanAkel, @zyj8881357: FYI, I think the topo file for the atmospheric model is smoothed, so even if the topo data underpinning the Pfafstetter watershed delineations and the atm topo data are both based on SRTM, I wouldn't expect a perfect match.  And then there's the fact that SRTM is limited to ~60N/S and high-lat topo is from some other source.
@biljanaorescanin

**Comment by biljanaorescanin** [comment]
Santha each file has .bin version and .nc4 

/discover/nobackup/projects/gmao/bcs_shared/legacy_bcs/Icarus-NLv3/NCAR_TOPO_GMTED/TOPO_CF0720x6C/smoothed/gmted_DYN_ave_720x4320.nc4 this is one we link into topo_DYN_ave_720x4320.data

**Comment by zyj8881357** [comment]
> > Hi Santha, I think you want me to compare the elevation diff between our land source data and the DEM from the link? I'd like to do the comparison, but the GEOS5_10s_mask.nc only provides the land-sea mask and does not include the elevation values. So I may need to get the elevation source data in our land model first before doing the comparison. @gmao-rreichle @biljanaorescanin
> 
> @zyj8881357 I recall, sometime ago when I plotted `GEOS5_10arcsec_mask_freshwater-lakes.nc`
> 
> https://github.com/GEOS-ESM/GEOSgcm_GridComp/blob/5938f19b7acfb46ea44ad42f421448ee03762241/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/Raster/makebcs/make_bcs#L437
> 
> it had elevation over land and none (=0) over the open ocean. Maybe I'm wrong (I do forget things)! But, yes, elevation diff is what I'm looking for ...
> 
> I'll also look at `${GLOBAL_CATCH_DATA}` that I'm feeding to the `make_bcs`. Information [such as this](https://portal.opentopography.org/datasetMetadata?otCollectionID=OT.122019.4326.1) has made me curious about any differences!
> 
> Thanks!

Hi Santha, here is the comparison between the elevations used in the land model (srtm30_withKMS_2.5x2.5min.nc) and from your link (SRTM15+). While there is no substantial difference between them (panels a and b), but small random differences occur outside of 60S-60N (panel c). @sanAkel @biljanaorescanin @gmao-rreichle 

<img width="921" alt="image" src="https://user-images.githubusercontent.com/32678746/233462402-14ff3b9e-a842-4d40-99eb-d209c5e95c86.png">

**Comment by sanAkel** [comment]
Thank you @zyj8881357 for this update.

Locations (such as the Canadian coast) in the Arctic, Greenland, Antarctic, Western Pacific, they all seem to have differences- in other words, _uncertainty_ in coastline is evident. 

I always knew of such problematic spots when comparing our land-model mask with that from the ocean (from GEBCO).
Your plots clearly shows that we can trace those differences to the source:
- Land model (probably some version of SRTM?).
- Ocean model (GEBCO 2021).

**Comment by gmao-rreichle** [comment]
@sanAkel, @zyj8881357 : The GEOS catchments were defined (and mean elevations derived) using SRTM data where available, ie., ~60S-60N.  Outside of these lats, SRTM did not provide data.  (The shuttle didn't fly poleward of 60 N/S lat.)  I don't know what the "link to SRTM15+" is, but I suspect that the data at this link were backfilled outside of 60 N/S differently from what was done in the Raster processing.  I'm not sure what this is comparison is meant to accomplish.  I don't see a cause for concern.
cc: @rdkoster @biljanaorescanin

**Comment by lcandre2** [comment]
> > Thanks for the tag @gmao-rreichle. @zyj8881357, we do not currently have pfastetter codes for Greenland or Antarctica because most runoff flows through the ice sheet and then along the ice sheet bed to a land or ocean outlet. I have a routing code that addresses this and provides better output locations for greenland and antarctica (in the rare cases it has it) runoff. However, it is offline. @zyj8881357, can we touch base via a teams meeting to coordinate how to approach this? It might be the right time to develop catchments for the ice sheets. We have the data, I just need to do the work.
> 

All, @zyj8881357, @biljanaorescanin, apologies for the long delay. I have produced outlets and associated drainage basins for Greenland that are gridded to the GEOS510arcsec_mask.nc. The steps closely, but not completely follow that of Verdin (2013, Final Report: High Resolution Topographic Analysis for GMAO’s Catchment LSM) and use ArcticDEM v3 100m mosaic cropped to Greenland (Porter et al., 2018; https://www.pgc.umn.edu/data/arcticdem/). 100 m is approximately 10 arc seconds at a latitude of 70° N. Steps 3-6 were completed using Topotoolbox v2 (Schwanghart & Scherler, 2014; https://topotoolbox.wordpress.com/). 

1.     Crop DEM Greenland only. 
2.     Fill a small number of missing grid cells using bilinear interpolation. No extrapolation was used, and this step was only designed to remove NaN values assigned to a handful of locations across Greenland.
4.     Hydrologically condition DEM by filling sinks.
5.     Calculate flow direction and flow accumulation grids using the D8 routing algorithm.
6.     Delineate ‘surface streams’ and drainage basins using an area threshold of 250 km2, the same threshold as in Verdin (2013). Streams are extracted to vector format and contain the Strahler stream order numbering system; drainage basins retain a raster format. See below for streams and 
7.     Extract drainage basin outlets. For Greenland, at the given resolution there are 525 outlet locations.
8.     Map drainage basins and outlets to the GEOS10arcsec_mask.nc grid using ‘nearest’. This mapping follows the methodology outlined in Verdin (2013) & Mahanama et al. (2015; GMAO Tech Report 39).
9.     Confirm each point is appropriately assigned a drainage basin number. 
 
Notes (in no particular order)
1.     @zyj8881357 expect an email about this and about how you would like the data transferred shortly.
2.     I have done this for all of Greenland (both ice and land) because Yujin indicated a lack of outlets across Greenland and because previous efforts excluded Greenland. I can also produce drainage basins and outlets for ice (or ice sheet) only points as well. This would reduce the number of outlet locations around Greenland by excluding many of the islands.
3.     Outlet locations are on land and still need to be moved to the nearest ocean cell.
4.     This information should be sufficient to create the routing table. While Pfaffstetter interbasins can be determined using the surface streams from step 5, including Greenland ice sheet Pfaffstetter coding system would likely cause issues (or require a rearrangement of the tile numbering) in subsequent processes because the 200000000 ice value would be replaced. My suggestion would be to append add a basin number to the Greenland ice values instead – though this is completely up for discussion because I do not know what the exact impacts would be.
5.     In the original routing scheme, is that water from land ice (200000000) locations is routed to the nearest ocean location using an expanding spiral search, so some modification of the code will be needed to exclude the Greenland ice points because my previous work modified the routing table after the spiral search.
6.     There should be future consideration of how to deal with glacier runoff routing. 
7.     Because there is no dynamic ice sheet, water is theoretically routed across the surface. In reality, much of the water produced on the ice surface only flows short distances before flowing englacially to the ice sheet bed. The traditional routing scheme for subglacial water accounts for the overburden pressure of the ice and basal topography in the potential gradient calculation; however, to substantially differ from surface routing the bed slope must exceed the surface slope by ~10x, which is uncommon. 
8.     Though the streams and drainage basins reach the ice sheet divide, runoff generally only occurs along the ice sheet edges. 




![Picture1](https://github.com/GEOS-ESM/GEOSgcm_GridComp/assets/53305329/06a5159e-441f-4304-82f4-257a4d9b60a4)

Note: middle panel legend refers to stream order.

**Comment by sanAkel** [comment]
> > > Thanks for the tag @gmao-rreichle. @zyj8881357, we do not currently have pfastetter codes for Greenland or Antarctica because most runoff flows through the ice sheet and then along the ice sheet bed to a land or ocean outlet. I have a routing code that addresses this and provides better output locations for greenland and antarctica (in the rare cases it has it) runoff. However, it is offline. @zyj8881357, can we touch base via a teams meeting to coordinate how to approach this? It might be the right time to develop catchments for the ice sheets. We have the data, I just need to do the work.
> 
> All, @zyj8881357, @biljanaorescanin, apologies for the long delay. I have produced outlets and associated drainage basins for Greenland that are gridded to the GEOS510arcsec_mask.nc. The steps closely, but not completely follow that of Verdin (2013, Final Report: High Resolution Topographic Analysis for GMAO’s Catchment LSM) and use ArcticDEM v3 100m mosaic cropped to Greenland (Porter et al., 2018; https://www.pgc.umn.edu/data/arcticdem/). 100 m is approximately 10 arc seconds at a latitude of 70° N. Steps 3-6 were completed using Topotoolbox v2 (Schwanghart & Scherler, 2014; https://topotoolbox.wordpress.com/).
> 
> 1.     Crop DEM Greenland only. 2.     Fill a small number of missing grid cells using bilinear interpolation. No extrapolation was used, and this step was only designed to remove NaN values assigned to a handful of locations across Greenland. 4.     Hydrologically condition DEM by filling sinks. 5.     Calculate flow direction and flow accumulation grids using the D8 routing algorithm. 6.     Delineate ‘surface streams’ and drainage basins using an area threshold of 250 km2, the same threshold as in Verdin (2013). Streams are extracted to vector format and contain the Strahler stream order numbering system; drainage basins retain a raster format. See below for streams and 7.     Extract drainage basin outlets. For Greenland, at the given resolution there are 525 outlet locations. 8.     Map drainage basins and outlets to the GEOS10arcsec_mask.nc grid using ‘nearest’. This mapping follows the methodology outlined in Verdin (2013) & Mahanama et al. (2015; GMAO Tech Report 39). 9.     Confirm each point is appropriately assigned a drainage basin number.   Notes (in no particular order) 1.     @zyj8881357 expect an email about this and about how you would like the data transferred shortly. 2.     I have done this for all of Greenland (both ice and land) because Yujin indicated a lack of outlets across Greenland and because previous efforts excluded Greenland. I can also produce drainage basins and outlets for ice (or ice sheet) only points as well. This would reduce the number of outlet locations around Greenland by excluding many of the islands. 3.     Outlet locations are on land and still need to be moved to the nearest ocean cell. 4.     This information should be sufficient to create the routing table. While Pfaffstetter interbasins can be determined using the surface streams from step 5, including Greenland ice sheet Pfaffstetter coding system would likely cause issues (or require a rearrangement of the tile numbering) in subsequent processes because the 200000000 ice value would be replaced. My suggestion would be to append add a basin number to the Greenland ice values instead – though this is completely up for discussion because I do not know what the exact impacts would be. 5.     In the original routing scheme, is that water from land ice (200000000) locations is routed to the nearest ocean location using an expanding spiral search, so some modification of the code will be needed to exclude the Greenland ice points because my previous work modified the routing table after the spiral search. 6.     There should be future consideration of how to deal with glacier runoff routing. 7.     Because there is no dynamic ice sheet, water is theoretically routed across the surface. In reality, much of the water produced on the ice surface only flows short distances before flowing englacially to the ice sheet bed. The traditional routing scheme for subglacial water accounts for the overburden pressure of the ice and basal topography in the potential gradient calculation; however, to substantially differ from surface routing the bed slope must exceed the surface slope by ~10x, which is uncommon. 8.     Though the streams and drainage basins reach the ice sheet divide, runoff generally only occurs along the ice sheet edges.
> 
> ![Picture1](https://user-images.githubusercontent.com/53305329/241272863-06a5159e-441f-4304-82f4-257a4d9b60a4.png)
> 
> Note: middle panel legend refers to stream order.

Fantastic! Thank you @lcandre2 can you please push your script(s) for steps: `1- 9` to github or somewhere else?

**Comment by gmao-rreichle** [comment]
Many thanks, @lcandre2!!  Re. the scripts, they should definitely be added to github but we need to coordinate where exactly they should go.

**Comment by lcandre2** [comment]
Agreed @gmao-rreichle and @sanAkel. I have a bit of commenting to clean up, but I am open to where these should be located. I was thinking somewhere like
@GEOSgcm_GridComp/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils
But, I am open to all other suggestions too.

**Comment by sanAkel** [comment]
> Agreed @gmao-rreichle and @sanAkel. I have a bit of commenting to clean up, but I am open to where these should be located. I was thinking somewhere like
> 
> @GEOSgcm_GridComp/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils
> 
> But, I am open to all other suggestions too.

Great! Thanks @lcandre2.

While I have particular preference to following, just listing our choices:
1. Your ⬆️ suggested path
2. This repo: https://github.com/GEOS-ESM/GEOS_Util
3. https://github.com/GEOS-ESM/GMAO_Shared/tree/main/GEOS_Shared

If above choices: 2 and 3 are to be considered, then it would a be under a new folder.

I don't think there are any existing code depends on _this_, because it's new!

For that reason, and to make @sdrabenh's life simpler, I lean towards 2 ⬆️.

**Comment by lcandre2** [comment]
> @lcandre2, @sanAkel: If I'm not mistaken, the code in question is a tool to create input for make_bcs. We do have similar code for generating land surface inputs to make_bcs here: `GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/Raster/preproc/` I suggest placing @lcandre2's code into a new sub-directory `./route/landice/` FYI, there is a lot more such code that needs to be captured. We're trying to organize our work on make_bcs here: https://github.com/GEOS-ESM/GEOSgcm_GridComp/projects/4 The "preproc" element is captured under item 6 in the current To-Do list: "Identify ancillary input files to make_bcs that require preprocessing..." (The To-Do list is in need of updating and should change frequently.)
> 
> cc: @biljanaorescanin @weiyuan-jiang

@sanAkel , @gmao-rreichle , I think Rolf's approach is the most logical to me since these scripts are directly related to the land surface. I'll clean them up and make a pull request with the additions.

**Comment by zyj8881357** [comment]
Hi all,

For this issue, I have a new version of the outlets locations. I have emailed to @yvikhlya about this, and also make an update here.

For the new version of the outlets locations files, the updates are:
1. Outlets from Greenland created by Lauren are included.
2. The locations are represented by the lat lon indexes for the grid CF0180x6C_TM1440xTM1080-Pfafstetter.rst (43200x21600), as identified in this issue.
3. All the locations were moved to the ocean cells prescribed by the rst file CF0180x6C_TM1440xTM1080-Pfafstetter.rst .
4. Endorheic basins (with inland sinks) are not included in the files to ensure that all the outlet locations are at ocean cells. This issue should be treated in the future for the water balance issue.

Anyone can download them at:
https://drive.google.com/drive/folders/1a2eS6zaAFRVVSpms7WqJbFrmrxEITVCD?usp=sharing

Three files are available:
1. latraster.nc: The file shows the lat indexes of the outlets for each cell in CF0180x6C_TM1440xTM1080-Pfafstetter.rst
2. lonraster.nc : The file shows the lon indexes of the outlets for each cell in CF0180x6C_TM1440xTM1080-Pfafstetter.rst
3. Outlet_latlon.43200x21600 : A outlet location files (unformatted) that could be directly read by make_bcs (specifically, https://github.com/GEOS-ESM/GEOSgcm_GridComp/blob/develop/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/Raster/makebcs/mk_runofftbl.F90).
This file is generated from latraster.nc and lonraster.nc through my read_riveroutlet.F90 (also uploaded in the package). I generated the file for your convenience. But if any change needs to be made for the locations, they should be made on the latraster.nc and lonraster.nc, and then get the Outlet_latlon.43200x21600 regenerated .

I hope this can help to fix the issue. In the future, improvements could be made on the file mk_runofftbl.F90 to let it be able to read the locations in degree (rather than in indexes as the current version) and adjust those locations based on different kinds of ocean grids.

**Comment by yvikhlya** [comment]
@zyj8881357 

> All the locations were moved to the ocean cells prescribed by the rst file CF0180x6C_TM1440xTM1080-Pfafstetter.rst 

Does this mean MOM ocean cells or GEOS ocean cells? Do all *.TRN files for coupled grids need to be regenerated now?

**Comment by zyj8881357** [comment]
> @zyj8881357
> 
> > All the locations were moved to the ocean cells prescribed by the rst file CF0180x6C_TM1440xTM1080-Pfafstetter.rst
> 
> Does this mean MOM ocean cells or GEOS ocean cells? Do all *.TRN files for coupled grids need to be regenerated now?

I think they are in the MOM ocean cells now, but you may need to check this to see if further adjustments are needed. 
Yes, you need to use the file Outlet_latlon.43200x21600 to regenerate the *.TRN files by the make_bcs.

**Comment by yvikhlya** [comment]
No, this does not take into account ocean land-sea mask. Land-sea  mask adjustment is better to do in mk_runofftbl.F90, which is still missing this part.

**Comment by yvikhlya** [comment]
@zyj8881357 There is a little inconsistency in both latraster.nc and lonraster.nc. In both files `data` field has attribute `_FillValue = -9999`, but the actual sentinel value is -999. This confuses netcdf tools.

**Comment by zyj8881357** [comment]
> @zyj8881357 There is a little inconsistency in both latraster.nc and lonraster.nc. In both files `data` field has attribute `_FillValue = -9999`, but the actual sentinel value is -999. This confuses netcdf tools.

Sorry for the inconvenience. Yes, the actual missing value is set to -999 to be consistent with the value used in the mk_runofftbl.F90. But I forgot to change it in the netcdf attribution. Do you need me to fix that for the nc files, or you can do it by yourself?

**Comment by yvikhlya** [comment]
> > @zyj8881357 There is a little inconsistency in both latraster.nc and lonraster.nc. In both files `data` field has attribute `_FillValue = -9999`, but the actual sentinel value is -999. This confuses netcdf tools.
> 
> Sorry for the inconvenience. Yes, the actual missing value is set to -999 to be consistent with the value used in the mk_runofftbl.F90. But I forgot to change it in the netcdf attribution. Do you need me to fix that for the nc files, or you can do it by yourself?

I can fix these files if I need to use them. Just pointing out inconsistency in case anybody else wants to use them.

**Comment by sanAkel** [comment]
> > @zyj8881357 thank you!
> > Can you upload an ascii file with:
> > River (name) 	lat (deg N) 	lon (deg E)
> > ABCD 	y_1 	x_1
> > You should be able to upload it _here_, if not, please share via google drive; assuming no access to discover cluster at NCCS.
> > Thanks again.
> 
> I uploaded it to the google drive. The link is: https://docs.google.com/spreadsheets/d/1FbQsqoYL-knSmF_qzKlOcC8Pp3kxW4Ok/edit?usp=share_link&ouid=110154394974887000797&rtpof=true&sd=true
> 
> It's a spreadsheet. I also added the mean discharge for each river. Please check whether these locations are consistent with the ocean mask in the ocean model.

Hi @zyj8881357, Would you be able to upload this and share a link? ⬆️ [link](https://docs.google.com/spreadsheets/d/1FbQsqoYL-knSmF_qzKlOcC8Pp3kxW4Ok/edit?usp=share_link&ouid=110154394974887000797&rtpof=true&sd=true) does not work.
If you prefer, you can [email me](santha.akella@noaa.gov). Thanks a lot!

**Comment by zyj8881357** [comment]
> > > @zyj8881357 thank you!
> > > Can you upload an ascii file with:
> > > River (name) 	lat (deg N) 	lon (deg E)
> > > ABCD 	y_1 	x_1
> > > You should be able to upload it _here_, if not, please share via google drive; assuming no access to discover cluster at NCCS.
> > > Thanks again.
> > 
> > 
> > I uploaded it to the google drive. The link is: https://docs.google.com/spreadsheets/d/1FbQsqoYL-knSmF_qzKlOcC8Pp3kxW4Ok/edit?usp=share_link&ouid=110154394974887000797&rtpof=true&sd=true
> > It's a spreadsheet. I also added the mean discharge for each river. Please check whether these locations are consistent with the ocean mask in the ocean model.
> 
> Hi @zyj8881357, Would you be able to upload this and share a link? ⬆️ [link](https://docs.google.com/spreadsheets/d/1FbQsqoYL-knSmF_qzKlOcC8Pp3kxW4Ok/edit?usp=share_link&ouid=110154394974887000797&rtpof=true&sd=true) does not work. If you prefer, you can [email me](santha.akella@noaa.gov). Thanks a lot!

The issue should be solved in GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/Raster/makebcs/mk_runofftbl.F90 by introducing a subroutine outlets_to_ocean(Gridname,lons,lats,nx,ny) which moves the outlets on land to the nearest ocean point. Are you still working on this and wanting to see those locations? If so, I can run the code and make the spreadsheet for you again. 

/mk_runofftbl.F90

**Comment by sanAkel** [comment]
@zyj8881357 in Jul, I transferred to NOAA, EMC. I'm interested in looking at where your catchment points are, if you happen to have that spreadsheet handy and willing to share, it would help. Thanks!
